src/composables/useNotifications.ts

```diff
@@ -0,0 +1,118 @@
+import { orderBy, uniqueId, merge } from 'lodash';
+import { computed, ref } from 'vue';
+
+import { TransactionStatus } from './useTransactions';
+
+const DEFAULT_NOTIFICATION_TIMEOUT = 15 * 1_000; // 15s
+
+export type Notification = {
+  id: string;
+  type?: 'success' | 'error' | 'warning' | 'info';
+  transactionMetadata?: {
+    id: string;
+    status: TransactionStatus;
+    explorerLink: string;
+  };
+  title?: string;
+  message: string;
+  autoCloseAfterMs?: number; // set 0 to disable
+  addedTime?: number;
+};
+
+export type NewNotification = Pick<
+  Notification,
+  'type' | 'message' | 'title' | 'autoCloseAfterMs' | 'transactionMetadata'
+>;
+
+type NotificationsMap = Record<string, Notification>;
+
+const notificationsState = ref<NotificationsMap>({});
+
+// METHODS
+function addNotification(newNotification: NewNotification) {
+  const notificationsMap = getNotifications();
+  const notificationId = uniqueId('notification');
+
+  attemptToRemoveExistingNotification(newNotification);
+
+  notificationsMap[notificationId] = {
+    ...newNotification,
+    id: notificationId,
+    type: newNotification.type ?? 'info',
+    addedTime: Date.now(),
+    autoCloseAfterMs:
+      newNotification.autoCloseAfterMs ?? DEFAULT_NOTIFICATION_TIMEOUT
+  };
+
+  setNotifications(notificationsMap);
+
+  return notificationId;
+}
+
+function attemptToRemoveExistingNotification(newNotification: NewNotification) {
+  if (newNotification.transactionMetadata?.id) {
+    const previousNotificationFound = notifications.value.find(
+      previousNotification =>
+        previousNotification.transactionMetadata?.id ===
+        newNotification.transactionMetadata?.id
+    );
+    if (previousNotificationFound != null) {
+      removeNotification(previousNotificationFound.id);
+    }
+  }
+}
+
+function getNotification(id: string) {
+  const notificationsMap = getNotifications();
+
+  return notificationsMap[id];
+}
+
+function clearAllNotifications() {
+  setNotifications({});
+}
+
+function updateNotification(id: string, updates: Partial<Notification>) {
+  const notificationsMap = getNotifications();
+  const notification = getNotification(id);
+
+  if (notification != null) {
+    notificationsMap[id] = merge(notification, updates);
+
+    setNotifications(notificationsMap);
+  }
+}
+
+function removeNotification(id: string) {
+  const notificationsMap = getNotifications();
+
+  delete notificationsMap[id];
+
+  setNotifications(notificationsMap);
+}
+
+function getNotifications() {
+  return notificationsState.value;
+}
+
+function setNotifications(notifications: NotificationsMap) {
+  notificationsState.value = notifications;
+}
+
+const notifications = computed(() =>
+  orderBy(Object.values(getNotifications()), 'addedTime', 'desc')
+);
+
+export default function useNotifications() {
+  return {
```

Comment:
@garethfuller @1saf this is a pretty good pattern I think... as seen in this instance I actually have no local functions/computed etc that need to be inside `useNotifications`. However, there might be some in the future which is why I still think it's good to export it like so. This makes the import of this composable not different from any other... and there is a big performance gain since all these functions are not redeclared each time it's consumed.

I think we should go over all our composables and put as much of it as we can outside the composable.